lark/exceptions.py

```diff
@@ -116,44 +118,71 @@ def match_examples(self, parse_fn, examples, token_type_match_fallback=False, us
 
         return candidate[0]
 
+    def _format_terminals(self, terminals):
+        if self._terminals_by_name:
+            ts = []
+            for ter in terminals:
+                ts.append(self._terminals_by_name[ter].user_repr())
+        else:
+            ts = terminals
+        return "Expected one of: \n\t* %s\n" % '\n\t* '.join(ts)
+
+
 class UnexpectedEOF(ParseError, UnexpectedInput):
-    def __init__(self, expected, state=None):
+    def __init__(self, expected, state=None, _terminals_by_name=None):
         self.expected = expected
         self.state = state
         from .lexer import Token
-        self.token = Token("<EOF>", "") #, line=-1, column=-1, pos_in_stream=-1)
+        self.token = Token("<EOF>", "")  # , line=-1, column=-1, pos_in_stream=-1)
         self.pos_in_stream = -1
         self.line = -1
         self.column = -1
+        self._terminals_by_name = _terminals_by_name
+
+        super(UnexpectedEOF, self).__init__()
 
-        message = ("Unexpected end-of-input. Expected one of: \n\t* %s\n" % '\n\t* '.join(x.name for x in self.expected))
-        super(UnexpectedEOF, self).__init__(message)
+    def __str__(self):
+        # Be aware: Broken __str__ for Exceptions are terrible to debug. Make sure there is as little room as possible for errors
```

Comment:
I gave it some thought since last time, and I don't think Lark is the right place for general Python tips.

Reply:
Sure